test/arcade/core/util/ColorsTest.java

```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Test
+    public void getColor_calledInvalidNumber_returnsEmptyColor() {
+        Colors colors = new Colors(new Color(1, 1, 1, 1), new Color(255, 255, 255, 1), 0.0, 1.0);
+        assertEquals(new Color(0, 0, 0, 0), colors.getColor(-1.0));
+        assertEquals(new Color(0, 0, 0, 0), colors.getColor(-1.5));
+        assertEquals(new Color(0, 0, 0, 0), colors.getColor(-2.0));
+    }
+
+    @Test
+    public void getColor_calledBetweenMinAndDefault_returnsMinColor()  {
+        Colors colors = new Colors(new Color(1, 1, 1, 1), new Color(255, 255, 255, 1), 0.0, 1.0);
+        assertEquals(new Color(1, 1, 1, 1), colors.getColor(-0.5));
+        assertEquals(new Color(1, 1, 1, 1), colors.getColor(-0.1));
+        assertEquals(new Color(1, 1, 1, 1), colors.getColor(0.0));
+    }
+
+    @Test
+    public void getColor_calledLargerThanMax_returnsMaxColor() {
+        Colors colors = new Colors(new Color(1, 1, 1, 1), new Color(200, 200, 200, 1), 0.0, 1.0);
+        assertEquals(new Color(200, 200, 200, 1), colors.getColor(1.0));
+        assertEquals(new Color(200, 200, 200, 1), colors.getColor(2.0));
+        assertEquals(new Color(200, 200, 200, 1), colors.getColor(1.5));
+    }
+
+    @Test
+    public void getColor_calledBetweenMinAndMax_returnsInterpolatedSingleColor() {
+        Colors colors = new Colors(new Color(1, 0, 0, 1), new Color(201, 0, 0, 1), 0.0, 1.0);
+        assertEquals(new Color(51, 0, 0, 1), colors.getColor(0.25));
+        assertEquals(new Color(101, 0, 0, 1), colors.getColor(0.5));
+        assertEquals(new Color(151, 0, 0, 1), colors.getColor(0.75));
+    }
+
+    @Test
+    public void getColor_calledBetweenMinAndMax_returnsInterpolatedColor() {
+        Colors colors = new Colors(new Color(1, 1, 1, 1), new Color(201, 201, 201, 1), 0.0, 1.0);
+        assertEquals(new Color(51, 51, 51, 1), colors.getColor(0.25));
+        assertEquals(new Color(101, 101, 101, 1), colors.getColor(0.5));
+        assertEquals(new Color(151, 151, 151, 1), colors.getColor(0.75));
+    }
+
+    @Test
+    public void getColor_calledWithMod_returnsCorrectColor() {
+        Colors colors = new Colors(new Color[] {
+            new Color(0, 0, 0, 0),
+            new Color(100, 100, 100, 0),
+            new Color(200, 200, 200, 200),
+        }, 3);
+        assertEquals(new Color(0, 0, 0, 0), colors.getColor(3));
+        assertEquals(new Color(100, 100, 100, 0), colors.getColor(4));
+        assertEquals(new Color(200, 200, 200, 200), colors.getColor(5));
+    }
+
+    @Test
+    public void getColor_NonlinearColorsCalledBetweenMinAndMax_returnsInterpolatedColor() {
+        Colors colors = new Colors(new Color[] {
+            new Color(0, 0, 0, 0),
+            new Color(100, 100, 100, 0),
+            new Color(200, 200, 200, 200),
+        }, new double[] { 0, 0.1, 1});
+
+        assertEquals(new Color(0, 0, 0, 0), colors.getColor(0));
+        // assertEquals(new Color(50, 50, 50, 0), colors.getColor(0.05));
```

Comment:
I think this is desired functionality but it's low priority and the result is close - should I leave this as a commented out line or should the line be deleted?

Reply:
The weirdness is because "continuous" colormaps are pre-interpolated to a constant number of `BINS` (then treated like a lookup table) which gets you some rounding errors. The following works by producing a 256-length colormap with the color matching the index (but you need to set BINS = 255 or update the `Colors` constructor to compensate for the `-1` -- this breaks a couple of the other tests).

```java
@Test
  public void getColor_NonlinearColorsCalledBetweenMinAndMax_returnsInterpolatedColor() {
    Colors colors =
            new Colors(
                    new Color[] {
                        new Color(0, 0, 0, 0),
                        new Color(64, 64, 64, 64),
                        new Color(255, 255, 255, 255),
                    },
                    new double[] {0, 64. / 255, 1});

    assertEquals(new Color(0, 0, 0, 0), colors.getColor(0));
    assertEquals(new Color(4, 4, 4, 4), colors.getColor(4. / 255));
    assertEquals(new Color(64, 64, 64, 64), colors.getColor(64. / 255));
    assertEquals(new Color(192, 192, 192, 192), colors.getColor(192. / 255));
    assertEquals(new Color(255, 255, 255, 255), colors.getColor(1));
}
```

Reply:
I'm not sure this is testing the same thing here? The colors are still linearly mapped from 0 - 1 here (which the linear interpolation works generally). I was trying to test another aspect where the the color map isn't linear between the 3 numbers that are specified in the double array.

Reply:
Ah I see. I thought the idea was to test nonlinear spacing in the doubles array, and we can be clever about the associated colors so we don't run into rounding errors.

If you want to specifically test different spacing the colormap, I think you just need to work out the numbers/colors so that they divide nicely into the 256 bins.

Reply:
the analogous implementation here that I was testing was from the following in `PottsColorMaps.java` 

```java
    /** Color map for cell cytoplasm region. */
    public static final Colors MAP_CYTOPLASM =
            new Colors(
                    new Color[] {
                        new Color(0, 0, 0, 0),
                        new Color(255, 0, 255, 50),
                        new Color(255, 0, 255, 255),
                    },
                    new double[] {0, 0.1, 1});
```

From the test, the functionality does give similar results: 

When testing: `assertEquals(new Color(50, 50, 50, 0), colors.getColor(0.05));`, the result is `Color(48, 48, 48, 0)`. 

It's likely due to a difference in how the BINS get allocated between the numbers, but testing for a close enough int does not feel correct here. I removed one instance of `round` (because the int cast already truncates, which is the expected behavior in the lookup table) to pass the test, but I am not sure the best way to handle this discrepancy since it is not a major issue and low priority.